### ISSUE TITLE
astropy table does not return a boolean value

### ISSUE THREAD

**Issue by kslong** [astropy/astropy#12826]
### Description

When reading back an astropy table that had been written in fixed_width_two_line format and had contained a boolian column, the boolean values get read  back as a string

### Expected behavior

I expected to get the table back in the formats in which it was writen

### Actual behavior

As noted, I got back a string value for True or False, and so I could not run the same code on the table that I could run before it was written out.

### Steps to Reproduce

1. [First Step]
2. [Second Step]
3. [and so on...]

```
from astropy.table import Table
import numpy as np

z=np.linspace(0,50)
x=Table([z,z],names=['foo','goo'])
x['Good']=True
x.info()
x.write('foo.txt',format='ascii.fixed_width_two_line',overwrite=True)
xx=ascii.read('foo.txt')
xx.info()
        

```

Here is what this produces

```
from astropy.table import Table

import numpy as np

​

z=np.linspace(0,50)

x=Table([z,z],names=['foo','goo'])

x['Good']=True

x.info()

x.write('foo.txt',format='ascii.fixed_width_two_line',overwrite=True)

xx=ascii.read('foo.txt')

xx.info()

        

<Table length=50>
name  dtype 
---- -------
 foo float64
 goo float64
Good    bool
<Table length=50>
name  dtype 
---- -------
 foo float64
 goo float64
Good    str4

```

### System Details

>>> import platform; print(platform.platform())
Darwin-19.6.0-x86_64-i386-64bit
>>> import sys; print("Python", sys.version)
Python 3.7.11 (default, Jul 27 2021, 07:03:16) 
[Clang 10.0.0 ]
>>> import erfa; print("pyerfa", erfa.__version__)
pyerfa 1.7.1.1
>>> import astropy; print("astropy", astropy.__version__)
astropy 4.2
>>> import scipy; print("Scipy", scipy.__version__)
Scipy 1.4.1
>>> import matplotlib; print("Matplotlib", matplotlib.__version__)
Matplotlib 3.5.0
>>>

**Comment by pllim** [comment]
A little clunky but seems to work for me when you specify a converter:

```python
>>> from astropy.table import Table
>>> xx = Table.read('foo.txt', format='ascii.fixed_width_two_line', converters={'Good': [ascii.convert_numpy(bool)]})
>>> xx.info()
<Table length=50>
name  dtype
---- -------
 foo float64
 goo float64
Good    bool
>>>  type(xx['Good'][0])
numpy.bool_
>>> xx['Good'][0] is True  # Not exactly Python native
False
>>> xx['Good'][0] == True
True
```

Maybe @taldcroft , @hamogu , or @dhomeier have better ideas.

**Comment by hamogu** [comment]
The problem is that there is a loss of information. When you write out the table, the writer knows that it's a bool value, but once written, there is no way in this format to know is that was initially a bool or a string. If you want to preserve the format of a column, you need to use a format that saves that metadata in some way, e.g. ascii.csv or fits.

As it is now @pllim 's suggestion is exactly what I would do.

That said, we *could* add a guessing step that says "if all the strings look like "true" or "false", then convert to a bool, similar to "if it all looks like a number, convert to an int. Thinking about it, that's what the code should do, so we should fix that.

**Comment by taldcroft** [comment]
I think the better answer is using ECSV, which does round-trip boolean values, or the maybe-soon-available fixed-width ECSV #12821 if you want human-readability as well.

At this point I'm pretty hesitant about changing the behavior of `io.ascii` in such a fundamental way after 10 years. I'm sure there is code out there that reads a table and has lines like `mask = table['Good'] == 'True'` and that would break after the change.

**Comment by pllim** [comment]
@taldcroft , I did try with ECSV but "Good" reads back as `str` with the provided example above.

**Comment by pllim** [comment]
Oh, wait... maybe I typed it wrong. Seems to work now!

```python
from astropy.table import Table
import numpy as np

z=np.linspace(0,50)
x=Table([z,z],names=['foo','goo'])
x['Good']=True
x.write('foo.txt',format='ascii.ecsv',overwrite=True)
xx=Table.read('foo.txt', format='ascii.ecsv')
xx.info()
```
```
<Table length=50>
name  dtype 
---- -------
 foo float64
 goo float64
Good    bool
```

**Comment by kslong** [comment]
Obviously, one can avoid this problem pretty trivially once it is recognized.  One can simply replace True and False by 1 and 0.  That said I don't think this is what a naive user would expect.  

And frankly while I know that one can use more complicated formats for tables, fixed_width_two_line is IMHB the most readable, and most easily handled by my collaborators who regularly read tables I create into Excell or its linux/mac equivalents.

**Comment by pllim** [comment]
If you want Excel I/O, CSV or ECSV might actually work better but that is besides the point. If you want more advanced data format support, it is not unfair to use a different format that actually supports them. "bool" is just one of the cases, what if you also want to share arrays, etc? `fixed_width_two_line` won't be able to support all of that?

**Comment by taldcroft** [comment]
@kslong - reading and writing data to ASCII formats requires making choices for formats that do not specify the data type. This can lead to differences in expectations. While you expect a column that contains "True" and "False" to be unambiguously `bool`, there are others who would write a `str` column with "True" and "False" values and expect that come back as `str`.

Thus for formats like fixed width we have to make a choice. More than 10 years ago in the original `asciitable` package the choice was made to only automatically convert numeric values but not try to infer boolean values from values that appear boolean.

The infrastructure is all there to have made a different choice. Indeed it is very simple to change the global default (see below). But given the long heritage, multiple easy workarounds, and large user base (including production settings), I am quite hesitant to do this in the release.
```
In [11]: ascii.TableOutputter.default_converters.insert(2, ascii.convert_numpy(bool))

In [12]: from astropy.table import Table
    ...: import numpy as np
    ...: 
    ...: z=np.linspace(0,50)
    ...: x=Table([z,z],names=['foo','goo'])
    ...: x['Good']=True
    ...: x.info()
    ...: x.write('foo.txt',format='ascii.fixed_width_two_line',overwrite=True)
    ...: xx=ascii.read('foo.txt')
    ...: xx.info()
    ...: 
<Table length=50>
name  dtype 
---- -------
 foo float64
 goo float64
Good    bool
<Table length=50>
name  dtype 
---- -------
 foo float64
 goo float64
Good    bool
```

**Comment by taldcroft** [comment]
@kslong - I'd like to close this issue with a `wont-fix` (sorry) but replace with adding documentation for a workaround #12828. OK?

**Comment by hamogu** [comment]
@kslong Thank you again for bringing this up. We don't hear enough input from typical users in many cases, so please do bring up other points that you run across!

One thing we have learned in the past though, is that a "typical user" also does not like their code broken. Something that worked with astropy 0.4, astropy 1.0 ,... all the way up to astropy 5.1 will lead to an outcry [*] if it's suddenly broken in astropy 6.0. While we can argue in this case if this is a bug or a feature, I agree with @taldcroft that there are almost certainly people out there who do `table['colname'] == 'True'` somewhere in their code. Call it a work-around, but if we change this behavior, their code that was working before, suddenly will throw errors, or, even worse, silently give wrong results because their comparison may no select different results in the table.

That's different for a feature that was just recently introduced or that was broken before (say that gave you a blank plot instead of the plot it should display). This example, however, does not behave like you and I expect, but it does behave according to a well defined set of rules (convert to numbers, leave everything else as string).

*: I can't find the thread now, but pandas had a very recent example that went viral on twitter, so instead I can point only to this xkcd comic https://xkcd.com/1172/ and say, while the comic is of course humerous, the idea is based on real experiences.

This is one of those example (there are a few others) where we would probably do things again if we started from scratch, but since this is not new code, there is a trade-off between an inconvenience for new users vs an inconvenience for old users who may (or may not, be we can't know) rely on the existing code.

**Comment by taldcroft** [comment]
@kslong - astropy certainly strives to take account of user preferences, but as a stable project used by a very large community, we avoid making breaking API changes when possible. 

To make such a change, we'd want to see broad community consensus that (1) making this behavior change is desirable and (2) it is worth breaking some existing code. That kind of consensus would need to come from wide discussion, maybe on astropy-dev and astropy (and maybe Facebook though I don't use that). I'd be happy if you wanted to lead such a discussion and advocate for an API change.

(Crossed in flight with the comments from @hamogu).

**Comment by eerovaher** [comment]
Would it be a good idea to add a new configuration item that would control if the data is read as strings or if conversion is automatically attempted?

**Comment by eerovaher** [comment]
The main advantage of having a configuration item would be that the users could edit the configuration file just once, whereas the workarounds would have to be applied every time they are needed.

**Comment by taldcroft** [comment]
These days I'm less and less inclined to have configuration options that silently change the behavior of code. In other words when a line of code or function behaves fundamentally differently based on a config item, then that makes it hard for readers of the code to know what it will do.

Instead, even if it is a little less convenient, it is better to be explicit.

**Comment by taldcroft** [comment]
Overall the `converters` interface is clunky and there is at least one issue to make it better (which I would need to dig up). I think the better way forward is to improve that interface with this use case in mind.
